src/index.ts

```diff
@@ -132,79 +160,102 @@ export async function downloadEntityAndContentFiles(
+  // 1. get the hashes of the latest snapshots in the remote server, retry 10 times
+  const snapshots = await getSnapshots(
     components,
     options.contentServer,
     options.requestMaxRetries
   )
 
-  // 2. download the snapshot file if it contains deployments
-  //    in the range we are interested (>= genesisTimestamp)
-  if (hash && lastIncludedDeploymentTimestamp && lastIncludedDeploymentTimestamp > genesisTimestamp) {
-    try {
-      // 2.1. download the snapshot file if needed
-      await downloadFileWithRetries(
-        components,
-        hash,
-        options.tmpDownloadFolder,
-        [options.contentServer],
-        new Map(),
-        options.requestMaxRetries,
-        options.requestRetryWaitTime
-      )
+  for (const snapshot of snapshots) {
+    const { hash, lastIncludedDeploymentTimestamp, replacedSnapshotHashes } = snapshot
+    logs.debug('Snapshot found.', { contentServer: options.contentServer, hash: snapshot.hash })
+    // 2. for each snapshot, download the snapshot file if it contains deployments
+    //    in the range we are interested (>= genesisTimestamp)
+    const shouldStreamSnapshot =
+      hash &&
+      lastIncludedDeploymentTimestamp &&
+      lastIncludedDeploymentTimestamp > genesisTimestamp &&
+      await components.processedSnapshots.shouldStream(hash, replacedSnapshotHashes)
 
-      // 2.2. open the snapshot file and process line by line
-      const deploymentsInFile = processDeploymentsInFile(hash, components)
-      for await (const rawDeployment of deploymentsInFile) {
-        const deployment = coerceEntityDeployment(rawDeployment)
-        if (!deployment) continue
-        // selectively ignore deployments by localTimestamp
-        if (deployment.localTimestamp >= genesisTimestamp) {
-          components.metrics.increment('dcl_entities_deployments_processed_total', metricLabels)
-          yield deployment
-        }
-        // update greatest processed timestamp
-        if (deployment.localTimestamp > greatestProcessedTimestamp) {
-          greatestProcessedTimestamp = deployment.localTimestamp
+    if (shouldStreamSnapshot) {
+      try {
+        // 2.1. download the snapshot file if needed
+        await downloadFileWithRetries(
+          components,
+          hash,
+          options.tmpDownloadFolder,
+          [options.contentServer],
+          new Map(),
+          options.requestMaxRetries,
+          options.requestRetryWaitTime
+        )
+
+        // 2.2. open the snapshot file and process line by line
+        const deploymentsInFile = processDeploymentsInFile(hash, components)
+        components.processedSnapshots.startStreamOf(hash)
+        let numberOfStreamedEntities = 0
+        for await (const deployment of deploymentsInFile) {
+
+          const deploymentTimestamp = 'entityTimestamp' in deployment ? deployment.entityTimestamp : deployment.localTimestamp
+
+          if (deploymentTimestamp >= genesisTimestamp) {
```

Comment:
do you remember why is this? could we add a comment here?

Reply:
the genesisTimestamp is used in reconnections. So let's say you synchronize with a catalyst correctly up until NOW(1). then the remote catalyst gets restarted, that will trigger a re-synchronization from scratch locally. The genesisTimestamp may be set to NOW(1) to ignore all the history up until that moment.

The same logic should be applied for pointer-changes